### PR TITLE
Add String UDF ngrams

### PR DESCRIPTION
**Pull request by chanllen**
Add ngrams udf function and unit tests, description

Return N-grams for the string.  N-grams are the set of length-N sequences of items in a string; an "item" is a unit returned after splitting on the delimiter.

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/string.rst [hunk 1]
```diff
@@ -163,6 +163,19 @@ String Functions
...
+
+    Return ``n``-grams for the ``string`` splitted by empty string ' '
+
+.. function:: ngrams(string, n, delimiter) -> array<varchar>
+
+    Return ``n``-grams for the ``string`` splitted by ``delimiter``
+
+.. function:: ngrams(string, n, delimiter, glue) -> array<varchar>
+
+    Return ``n``-grams for the ``string`` splitted by ``delimiter`` and concat by ``glue``
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/functions/string.rst]: `concatenated`?

**Comment by highker** [comment]
How about this...
```
SELECT
    reduce(
        SEQUENCE(1, length('abcdefghigk') - 4 + 1),
        CAST(array[] AS array(varchar)),
        (s, x) -> s || substr('abcdefghigk', x, 4),
        s -> s
    )

**Comment by dain** [comment]
I'm confused about how someone would use this function.  Can you give some examples?

Additionally, I'm not sure we want to mix the `n-gram` generation and the splitting together since the splitting is so limited in this.  Instead, I think we should consider having ngram take an array so it can work with any type (including binary strings).  I'm thinking of something like `ngram(3, split(string, delimiter)) ==> array<array<varhar>>`, then you can join them (or not) how ever you want.

**Comment by chanllen** [comment]
@highker Thank you for reviewing my diff. The solution should work but I would consider it as a hacky way since it is not efficient and not clear to users. Instead if we use the UDF, it is much clear to the end users if they have the similar use case.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestNgramsFunction.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        assertFunction("ngrams('\u8B49\u8BC1\u8A3C\u8BC1\u7231', 2, '\u8BC1', '')", new ArrayType(createVarcharType(5)), ImmutableList.of("\u8B49\u8A3C", "\u8A3C\u7231"));
+
+        // Test delimiter is empty
+        assertFunction("ngrams('foo', 1, '', '')", new ArrayType(createVarcharType(3)), ImmutableList.of("f", "o", "o"));
+        assertFunction("ngrams('foobar', 4, '', '')", new ArrayType(createVarcharType(6)), ImmutableList.of("foob", "ooba", "obar"));
+        assertFunction("ngrams('foobar', 4, '', '|')", new ArrayType(createVarcharType(6)), ImmutableList.of("f|o|o|b", "o|o|b|a", "o|b|a|r"));
+
+        assertInvalidFunction("ngrams('foo', 0)", "N must be positive");
+        assertInvalidFunction("ngrams('foo', -1)", "N must be positive");
+        assertInvalidFunction("ngrams('foo', 2147483648)", "N is too large");
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestNgramsFunction.java]: I love all these test cases.  Does it make sense to have a test case to cover empty string input?

File: presto-docs/src/main/sphinx/functions/string.rst (hunk 1 above)
Comment (dain) [reply presto-docs/src/main/sphinx/functions/string.rst]: also replace `splitted` with `split`

Comment (dain) [reply presto-docs/src/main/sphinx/functions/string.rst]: and maybe use `split on`

**Comment by chanllen** [comment]
@dain There are a lot of use cases like that when we try to extract the some common pattern of the texts For example, when we deal with users' complaints, we would like to know what are the Top words/phrases they mentioned in their texts.

I think your proposed solution totally makes sense. However, I would assume in that way we would consume more memory/computing resources (correct me if I am wrong) to achieve the this result in separate steps while I would assume 99% of the real life use cases would be on 'varchar' instead of other data type. There is a trade-off here I think
Would like to hear more thoughts from you :)

**Comment by electrum** [comment]
It sounds like we are debating flexibility vs performance. Can you write a JMH benchmark for both approaches? That will gives us hard data to determine if performance is an issue.

File: presto-docs/src/main/sphinx/functions/string.rst [hunk 3]
```diff
@@ -163,6 +163,19 @@ String Functions
 
     Returns the stem of ``word`` in the ``lang`` language.
 
+.. function:: ngrams(string, n) -> array<varchar>
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/string.rst]: Please move these up so that the functions within the section are sorted alphabetically.

**Comment by dain** [comment]
@chanllen an optimizer can be written to manually fuse common operations.  We did this for some JSON functions.  BTW, for your example "Top words/phrases", I don't think we need to join back the ngram into a single string until the output phase (or not at all).

**Comment by chanllen** [comment]
@dain I see. The first part sounds great!  For the second part, I am not sure if I fully understand. How can we not join it back together to get the phrases out since the sequence of words matters?

**Comment by dain** [comment]
@chanllen I'm suggesting you can build a function with the signature `ngram(N, split(string, delimiter)) ==> array<array<varhar>>`, so an example usage would look like this: 
```
ngrams(ARRAY['foo', 'bar', 'baz'], 2) ==> ARRAY[ARRAY['foo', 'bar'], ARRAY['bar', 'baz']]
```
This would allow for any generic kind of splitting (e.g. `regex_split`), and the function can be generic so it would work with any type, making it much more powerful.

As for performance, I believe this function can be written using dictionary blocks, so no extra copies would be needed.

**Comment by chanllen** [comment]
@dain Thank you very much! It totally make senses to me! Will build that!

**Comment by dain** [comment]
Superseded by #11612

**Pull request closed without merging**